File: J9Ch/src/J_2_Binary_Search/Required_10/_62_Search_in_Rotated_Sorted_Array_Medium.java
package J_2_Binary_Search.Required_10;
import java.util.*;
/**62. Search in Rotated Sorted Array
 * Medium
 * Created by tianhuizhu on 6/27/17.
 */
public class _62_Search_in_Rotated_Sorted_Array_Medium {

    public class Solution {
        public int search(int[] A, int target) {
            if (A == null || A.length == 0) {
                return -1;
            }

            int start = 0;
            int end = A.length - 1;
            int mid;

            while (start + 1 < end) {
                mid = start + (end - start) / 2;
                if (A[mid] == target) {
                    return mid;
                }
                if (A[start] < A[mid]) {
                    // situation 1, red line
                    if (A[start] <= target && target <= A[mid]) {
                        end = mid;
                    } else {
                        start = mid;
                    }
                } else {
                    // situation 2, green line
                    if (A[mid] <= target && target <= A[end]) {
                        start = mid;
                    } else {
                        end = mid;
                    }
                }
            } // while

            if (A[start] == target) {
                return start;
            }
            if (A[end] == target) {
                return end;
            }
            return -1;
        }
    }
}

File: J9Ch/src/J_2_Binary_Search/Required_10/python_cpp/_62_Search_in_Rotated_Sorted_Array_Medium.cpp
/**62. Search in Rotated Sorted Array
 * Medium
 * Created by tianhuizhu on 6/27/17.
 */


#include <vector>

using namespace std;

class Solution {
    /**
     * param A : an integer ratated sorted array
     * param target :  an integer to be searched
     * return : an integer
     */
private:
int find(vector<int> &A, int l, int r, int target) {
        if (l > r) {
            return -1;
        }
        int idx = -1;
        if (A[l] <= A[r]) {
            int left = l, right = r, mid = 0;
            while (left <= right) {
                mid = (left + right) >> 1;
                if (A[mid] == target) {
                    idx = mid;
                    break;
                }
                else if (A[mid] > target) {
                    right = mid - 1;
                }
                else {
                    left = mid + 1;
                }
            }
        }
        else {
            int mid = (l + r) >> 1;
            if (A[mid] == target) {
                idx = mid;
            }
            else {
                idx = find(A, l, mid - 1, target);
                idx = idx == -1 ? find(A, mid + 1, r, target) : idx;
            }
        }
        return idx;
}
public:
    int search(vector<int> &A, int target) {
        // write your code here
        int n = A.size();
        return find(A,0,n-1,target);
    }
};